crates/thermocouple-converter/src/lib.rs

```
@@ -1,117 +1,131 @@
-        48.30222,
-        -1.646031,
-        0.05464731,
-        -0.0009650715,
-        0.000008802193,
-        -0.00000003110810,
-        0.0,
-        0.0,
-        0.0,
+    [                   // Coefficients for 20.644 <= voltage <= 54.886
+       -1.318058E+02,
+        4.830222E+01,
+       -1.646031E+00,
+        5.464731E-02,
+       -9.650715E-04,
+        8.802193E-06,
+       -3.110810E-08,
+        0.000000E+00,
+        0.000000E+00,
+        0.000000E+00,
     ],
 ];
 
-///function for power
-/// in: base: f64, exp: i32
-/// out: result: f64
-fn pow(base: f64, exp: i32) -> f64 {
-    let mut result = 1.0;
-    if exp < 0 {
-        result = 1.0 / pow(base, -exp);
-    } else {
-        for _ in 0..exp {
-            result *= base;
-        }
-    }
-    return result;
+/// Converts a 32-bit ADC reading to a temperature in celsius.
+pub fn adc_to_celsius(adc_reading: i32) -> f64 {
+    voltage_to_celsius(adc_to_voltage(adc_reading))
 }
 
-///function to calculate the conversion between voltage to celsius from the thermocoupler.
-///in: voltage: f64
-///out: celsius: f64
+/// Converts a 32-bit ADC reading to a voltage.
+pub fn adc_to_voltage(adc_reading: i32) -> f64 {
+    const REFERENCE_VOLTAGE: f64 = 5.0;
+    const MAX_ADC_VALUE: f64 = 4_294_967_296.0;
+    const V_SCALE: f64 = REFERENCE_VOLTAGE / MAX_ADC_VALUE; 
+
+    adc_reading as f64 * V_SCALE
+}
+
+/// Converts voltage to celsius for type K thermocouples.
 pub fn voltage_to_celsius(voltage: f64) -> f64 {
-    //define variables
+    return match voltage {
+        -5.891..=0.0 => calc_temp_exponential(voltage, &TYPE_K_COEF[0]),
+        0.0..=20.644 => calc_temp_exponential(voltage, &TYPE_K_COEF[1]),
+        20.644..=54.886 => calc_temp_exponential(voltage, &TYPE_K_COEF[2]),
+
+        // Insane temperature ranges that should never be reached.
+        // Hitting this is a strong indicator of a bug in the Argus system.
+        _ => panic!("T < -270 or T > 1372 celcius")
```

Thoughts on the panic here? Theoretically these temperatures are not attainable for Argus, however we could return some number like -1.0 as it was before my refactoring if you prefer.

---

I think this is more of an error than a panic. Yes you are totally correct this rocket should never get that hot :hot_pepper: but I am strongly against panics that are not in init functions. You can create a thermocouple conversion error type and then add this to hydra error manager so that we can capture it in the error manager and just call xyz.temp_convert()?; 
I also am not a fan of returning -1 because it can be really cryptic and we can just throw an error nicely

---

I also don't like returning -1, it was just the prior implementation. We can implement an error for this of course, and I pondered that, but was just concerned about potentially complicating the API when hitting this would indicate much larger issues. I guess we can just handle this or whatever, and if it hits an error, we trigger a fault state in the SM and re-calibrate or something?

---

Yeah exactly, something like this would be a fault and we could for example that that temperature sensor offline if it persists or we just outright take it offline.

---

This way we can reduce the restart time compared to a panic

---

Fair enough. Do you want to implement that quickly then? I probably won't have time until after the reading break.